src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1510,6 +1517,27 @@ DistPartitionRelationId(void)
 }
 
 
+/*
+ * This function returns oid of pg_dist_partition relation from catalog, it
+ * specifically does not use cache and designed to used in situations where
+ * reading from cache is not safe.
+ */
+static Oid
+DistPartitionRelationIdViaCatalog(void)
+{
+	const char *relationName = "pg_dist_partition";
+	Oid distPartitionRelationId = get_relname_relid(relationName, PG_CATALOG_NAMESPACE);
+
+	if (distPartitionRelationId == InvalidOid)
+	{
+		ereport(ERROR, (errmsg("cache lookup failed for %s, called to early?",
```

Comment:
I think the error messages should be re-worded. 

Also, I'd write another function which calls `get_relname_relid ` and does the check. Later, `DistPartitionRelationIdViaCatalog ` and `DistPartitionLogicalRelidIndexIdViaCatalog ` calls that function.